op-node/rollup/derive/payload_queue.go

```diff
@@ -0,0 +1,144 @@
+package derive
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type L1ReceiptsFetcher interface {
+	Fetch(ctx context.Context, blockHash common.Hash) (eth.L1Info, types.Transactions, types.Receipts, error)
+}
+
+type PayloadQueueOutput interface {
+	AddSafeAttributes(attributes *eth.PayloadAttributes)
+	SafeL2Head() eth.L2BlockRef
+	StageProgress
+}
+
+type taggedBatch struct {
+	batch  *BatchData
+	origin eth.L1BlockRef
+}
+
+type AttributesQueue struct {
+	log      log.Logger
+	config   *rollup.Config
+	dl       L1ReceiptsFetcher
+	next     PayloadQueueOutput
+	progress Progress
+	batches  []*taggedBatch
+}
+
+func NewAttributesQueue(log log.Logger, cfg *rollup.Config, l1Fetcher L1ReceiptsFetcher, next PayloadQueueOutput) *AttributesQueue {
+	return &AttributesQueue{
+		log:    log,
+		config: cfg,
+		dl:     l1Fetcher,
+		next:   next,
+	}
+}
+
+func (pq *AttributesQueue) AddBatch(batch *BatchData) {
+	pq.log.Info("Received next batch", "batch_epoch", batch.EpochNum, "batch_timestamp", batch.Timestamp, "txs", len(batch.Transactions))
+	pq.batches = append(pq.batches, &taggedBatch{batch: batch})
+}
+
+func (pq *AttributesQueue) Progress() Progress {
+	return pq.progress
+}
+
+func (pq *AttributesQueue) Step(ctx context.Context, outer Progress) error {
+	if changed, err := pq.progress.Update(outer); err != nil || changed {
+		return err
+	}
+	attr, err := pq.DeriveL2Inputs(ctx, pq.next.SafeL2Head())
+	if err != nil {
+		return err
+	}
+	pq.next.AddSafeAttributes(attr)
+	return nil
+}
+
+func (pq *AttributesQueue) ResetStep(ctx context.Context, l1Fetcher L1Fetcher) error {
+	pq.batches = pq.batches[:0]
+	// TODO: progress thing here
+	pq.progress = pq.next.Progress()
```

Comment:
this TODO looks fine. It's already correctly adopting the progress of the inner stage upon reset.